File: solutions/elixir/binary-search-tree/1/lib/binary_search_tree.ex
defmodule BinarySearchTree do
  defstruct data: nil, left: nil, right: nil

  @type bst_node :: %BinarySearchTree{data: any, left: bst_node | nil, right: bst_node | nil}

  @doc """
  Create a new Binary Search Tree with root's value as the given 'data'
  """
  @spec new(any) :: bst_node
  def new(data) do
    %BinarySearchTree{data: data}
  end

  @doc """
  Creates and inserts a node with its value as 'data' into the tree.
  """
  @spec insert(bst_node, any) :: bst_node
  def insert(nil, data), do: new(data)

  def insert(tree, data) when tree.data >= data do
    Map.update!(tree, :left, &insert(&1, data))
    # alternative methods
    # Map.update(tree, :left, nil, &insert(&1, data))
    # %BinarySearchTree{tree | left: insert(tree.left, data)}
    # Map.put(tree, :left, insert(tree.left, data))
  end

  def insert(tree, data) when tree.data < data do
    Map.update!(tree, :right, &insert(&1, data))
  end

  @doc """
  Traverses the Binary Search Tree in order and returns a list of each node's data.
  """
  @spec in_order(bst_node) :: [any]
  def in_order(nil), do: []

  def in_order(tree) do
    in_order(tree.left) ++ [tree.data] ++ in_order(tree.right)
  end
end

File: solutions/typescript/binary-search-tree/1/binary-search-tree.ts
interface BinarySearchTreeNode {
  data: number;
  left?: BinarySearchTreeNode;
  right?: BinarySearchTreeNode;
}

export class BinarySearchTree {
  #data?: BinarySearchTreeNode;

  constructor(data: unknown) {
    if (typeof data === 'number') {
      this.#data = { data: data as number };
    } else if (data instanceof BinarySearchTree) {
      this.#data = data.#data;
    } else if (data && typeof data === 'object' && 'data' in (data as object)) {
      this.#data = data as BinarySearchTreeNode;
    }
  }

  get data(): number | undefined {
    return this.#data?.data;
  }

  get left(): BinarySearchTree | undefined {
    return this.#data?.left ? new BinarySearchTree(this.#data.left) : undefined;
  }

  get right(): BinarySearchTree | undefined {
    return this.#data?.right ? new BinarySearchTree(this.#data.right) : undefined;
  }

  insert(item: number): BinarySearchTree {
    if (!this.#data) {
      this.#data = { data: item };
      return this;
    }
    let node = this.#data;
    while (true) {
      if (item <= node.data) {
        if (node.left) {
          node = node.left;
        } else {
          node.left = { data: item };
          break;
        }
      } else {
        if (node.right) {
          node = node.right;
        } else {
          node.right = { data: item };
          break;
        }
      }
    }
    return this;
  }

  each(callback: (data: unknown) => void): void {
    const traverse = (node?: BinarySearchTreeNode): void => {
      if (!node) return;
      traverse(node.left);
      callback(node.data);
      traverse(node.right);
    };
    traverse(this.#data);
  }
}